source/common/router/router.cc

```diff
@@ -328,6 +336,29 @@ void Filter::chargeUpstreamCode(Http::Code code,
   chargeUpstreamCode(response_status_code, *fake_response_headers, upstream_host, dropped);
 }
 
+void Filter::finalizeTimeoutHeaders() {
+  // if we have a deadline, update the timeout headers.
+  if (deadline_) {
+    // calculate how much time we have left
+    std::chrono::milliseconds expected_timeout =
+        std::chrono::duration_cast<std::chrono::milliseconds>(deadline_.value() -
+                                                              config_.timeSource().monotonicTime());
+    // this _should_ be positive
+    // TODO: what should be done if it is not positive? as i believe this can happen in rare
+    // cases.
+    ASSERT(expected_timeout.count() > 0, "expected_timeout is not positive");
```

Comment:
Yeah I think it's possible for this to be delayed. In this case though I think the timeout timer should fire so we can just return here? Make sure this case has coverage.